src/number_theory/bernoulli_polynomials.lean

```diff
@@ -0,0 +1,171 @@
+  apply sum_congr rfl,
+  rintros x hx,
+  rw mem_range_succ_iff at hx, rw [choose_symm hx, nat.sub_sub_self hx],
+end
+
+namespace bernoulli_poly
+
+/-
+### examples
+-/
+
+section examples
+
+@[simp] lemma bernoulli_poly_zero : bernoulli_poly 0 = 1 :=
+by simp [bernoulli_poly]
+
+@[simp] lemma bernoulli_poly_eval_zero (n : ℕ) : (bernoulli_poly n).eval 0 = bernoulli n :=
+begin
+ rw [bernoulli_poly, polynomial.eval_finset_sum, sum_range_succ],
+  have : ∑ (x : ℕ) in range n, bernoulli x * (n.choose x) * 0 ^ (n - x) = 0,
+  { apply sum_eq_zero (λ x hx, _),
+    have h : 0 < n - x := nat.sub_pos_of_lt (mem_range.1 hx),
+    simp [h] },
+  simp [this],
+end
+
+end examples
+
+@[simp] theorem sum_bernoulli_poly (n : ℕ) :
+  ∑ k in range (n + 1), ((n + 1).choose k : ℚ) • bernoulli_poly k =
+    polynomial.monomial n (n + 1 : ℚ) :=
+begin
+ simp_rw [bernoulli_poly_def, finset.smul_sum, finset.range_eq_Ico, ←finset.sum_Ico_Ico_comm,
+    finset.sum_Ico_eq_sum_range],
+  simp only [cast_succ, nat.add_sub_cancel_left, nat.sub_zero, zero_add, linear_map.map_add],
+  simp_rw [polynomial.smul_monomial, mul_comm (bernoulli _) _, smul_eq_mul, ←mul_assoc],
+  conv_lhs { apply_congr, skip, conv
+    { apply_congr, skip,
+      rw [nat.prime.choose_mul ((nat.le_sub_left_iff_add_le (mem_range_le H)).1 (mem_range_le H_1))
+        (le.intro rfl), add_comm x x_1, nat.add_sub_cancel, mul_assoc, mul_comm, ←smul_eq_mul,
+        ←polynomial.smul_monomial], },
+    rw [←sum_smul], },
+  rw sum_range_succ,
+  simp only [add_right_eq_self, cast_succ, mul_one, cast_one, cast_add, nat.add_sub_cancel_left,
+    choose_succ_self_right, one_smul, bernoulli_zero, sum_singleton, zero_add,
+    linear_map.map_add, range_one],
+  have f : ∀ x ∈ range n, 2 ≤ n + 1 - x,
+  { rintros x H, rw [mem_range] at H,
+    exact nat.le_sub_left_of_add_le (succ_le_succ H) },
+  apply sum_eq_zero (λ x hx, _),
+  rw [sum_bernoulli _ (f x hx), zero_smul],
+end
+
+open power_series
+variables {A : Type*} [integral_domain A] [algebra ℚ A] [char_zero A]
+
+theorem exp_bernoulli_poly (t : A) :
+  power_series.mk (λ n, polynomial.aeval t ((1 / nat.factorial n : ℚ) • bernoulli_poly n))
+    * (exp A - 1) = X * rescale t (exp A) :=
+begin
+--opening and splitting sums
+  ext, rw [coeff_mul, coeff_mul, nat.sum_antidiagonal_eq_sum_range_succ_mk,
+    nat.sum_antidiagonal_eq_sum_range_succ_mk],
+  simp only [coeff_mk, coeff_one, coeff_exp, ring_hom.id_apply, linear_map.map_sub, factorial,
+    rat.algebra_map_rat_rat],
+  rw sum_range_succ, -- the `succ` term is 0
+  have f : ∀ x ∈ range n, ite (n - x = 0) 1 0 = (0 : A),
+  { rintros x hx, split_ifs,
+    { exfalso, apply mem_range_sub_ne_zero hx h, }, refl, },
+  conv_lhs { congr, skip, apply_congr, skip, rw f x H, }, --making LHS sum clean
+  cases n, { simp only [one_div, alg_hom.map_smul, power_series.coeff_zero_eq_constant_coeff,
+    add_zero, polynomial.aeval_one, if_congr, mul_one, nat.nat_zero_eq_zero,
```

Comment:
It is useful to extract this out as a sublemma?

Reply:
I managed to remove the hypothesis `f` and its proof. The rest seems to be specific to the proof..